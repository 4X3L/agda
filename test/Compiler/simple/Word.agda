{-# OPTIONS -v treeless:20 #-}
module Word where

open import Agda.Builtin.Nat
open import Agda.Builtin.Word renaming (primWord64ToNat to toNat)
open import Agda.Builtin.Equality
open import Agda.Builtin.Unit
open import Agda.Builtin.IO
open import Agda.Builtin.FromNat
open import Agda.Builtin.String

instance
  NumWord : Number Word64
  Number.Constraint NumWord _ = ⊤
  fromNat {{NumWord}} n = primWord64FromNat n

natOp : (Nat → Nat → Nat) → Word64 → Word64 → Word64
natOp f a b = fromNat (f (toNat a) (toNat b))
{-# INLINE natOp #-}

addWord : Word64 → Word64 → Word64
addWord = natOp _+_

postulate
  showWord : Word64 → String
  putStrLn : String → IO ⊤
{-# FOREIGN GHC import qualified Data.Text.IO #-}
{-# COMPILE GHC showWord = Data.Text.pack . show #-}
{-# COMPILE GHC putStrLn = Data.Text.IO.putStrLn #-}

{-# COMPILE JS showWord = function(x) { return x.toString(); } #-}
{-# COMPILE JS putStrLn = function (x) { return function(cb) { process.stdout.write(x + "\n"); cb(0); }; } #-}

test : Word64
test = addWord 9223372036854775908 9223372036854775809

prf : test ≡ 101
prf = refl

main : IO ⊤
main = putStrLn (showWord test)
