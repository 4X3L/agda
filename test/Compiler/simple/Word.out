EXECUTED_PROGRAM

ret > ExitSuccess
out > { compiling Word.NumWord
out > Word.NumWord =
out >   Agda.Builtin.FromNat.recCon-NOT-PRINTED
out >     _ (λ a _ → Agda.Builtin.Word.primWord64FromNat a)
out > }
out > { compiling Word.natOp
out > Word.natOp =
out >   λ a b c →
out >     Agda.Builtin.Word.primWord64FromNat
out >       (a (Agda.Builtin.Word.toNat b) (Agda.Builtin.Word.toNat c))
out > }
out > { compiling Word.addWord
out > Word.addWord =
out >   λ a b →
out >     Agda.Builtin.Word.primWord64FromNat
out >       (Agda.Builtin.Word.toNat a + Agda.Builtin.Word.toNat b)
out > }
out > { compiling Word.test
out > Word.test =
out >   Word.addWord
out >     (Agda.Builtin.Word.primWord64FromNat 9223372036854775908)
out >     (Agda.Builtin.Word.primWord64FromNat 9223372036854775809)
out > }
out > { compiling Word.prf
out > Word.prf = _
out > }
out > { compiling Word.main
out > Word.main = Word.putStrLn (Word.showWord Word.test)
out > }
out > 101
out >
